src/writers/writer.cpp

```diff
@@ -8,255 +8,451 @@
+                                      std::vector<file>& files)
+{
+    std::queue<fs::path> paths;
+    paths.push(base_dir);
+
+    if (!make_dirs_(paths)) {
+        return false;
+    }
+
+    // create directories
+    for (auto i = dimensions.size() - 2; i >= 1; --i) {
+        const auto& dim = dimensions.at(i);
+        const auto n_shards = common::shards_along_dimension(dim);
+        CHECK(n_shards);
+
+        auto n_paths = paths.size();
+        for (auto j = 0; j < n_paths; ++j) {
+            const auto path = paths.front();
+            paths.pop();
+
+            for (auto k = 0; k < n_shards; ++k) {
+                paths.push(path / std::to_string(k));
+            }
+        }
+
+        if (!make_dirs_(paths)) {
+            return false;
+        }
+    }
+
+    // create files
+    {
+        const auto& dim = dimensions.front();
+        const auto n_shards = common::shards_along_dimension(dim);
+        CHECK(n_shards);
+
+        auto n_paths = paths.size();
+        for (auto i = 0; i < n_paths; ++i) {
+            const auto path = paths.front();
+            paths.pop();
+            for (auto j = 0; j < n_shards; ++j) {
+                paths.push(path / std::to_string(j));
+            }
+        }
+    }
 
-    return !failure;
+    return make_files_(paths, files);
 }
 
 bool
-zarr::FileCreator::create_c_dirs_(int n_c)
+zarr::FileCreator::make_dirs_(std::queue<fs::path>& dir_paths)
 {
-    std::latch latch(n_c);
-    std::atomic<bool> failure{ false };
-    for (auto c = 0; c < n_c; ++c) {
+    if (dir_paths.empty()) {
+        return true;
+    }
+
+    std::atomic<bool> success = true;
+
+    const auto n_dirs = dir_paths.size();
+    std::latch latch(n_dirs);
+
+    for (auto i = 0; i < n_dirs; ++i) {
+        const auto dirname = dir_paths.front();
+        dir_paths.pop();
+
         thread_pool_->push_to_job_queue(
-          std::move([base = base_dir_, c, &latch, &failure](std::string& err) {
+          [dirname, &latch, &success](std::string& err) -> bool {
               try {
-                  const auto path = base / std::to_string(c);
-                  if (fs::exists(path)) {
-                      EXPECT(fs::is_directory(path),
-                             "%s must be a directory.",
-                             path.c_str());
-                  } else if (!failure) {
-                      EXPECT(fs::create_directories(path),
-                             "Failed to create directory: %s",
-                             path.c_str());
+                  if (fs::exists(dirname)) {
+                      EXPECT(fs::is_directory(dirname),
+                             "'%s' exists but is not a directory.",
+                             dirname.c_str());
+                  } else if (success) {
+                      EXPECT(fs::create_directories(dirname),
+                             "Not creating directory '%s': another job failed.",
```

Comment:
How did you infer that another job failed in this case?

Reply:
That's actually incorrect. Will fix.